universe/src/storage/cassandra.rs

```diff
@@ -0,0 +1,407 @@
+use std::str::FromStr;
+
+use super::*;
+use proto::universe::KeyspaceInfo;
+use scylla::macros::{FromUserType, SerializeValue};
+use scylla::query::Query;
+use scylla::statement::SerialConsistency;
+use scylla::transport::errors::DbError;
+use scylla::transport::errors::QueryError;
+use scylla::transport::PagingState;
+use scylla::Session;
+use scylla::{FromRow, SerializeRow, SessionBuilder};
+use uuid::Uuid;
+
+pub struct Cassandra {
+    session: Session,
+}
+
+static CREATE_KEYSPACE_QUERY: &str = r#"
+    INSERT INTO chardonnay.keyspaces
+    (keyspace_id, name, namespace, primary_zone, base_key_ranges)
+    VALUES (?, ?, ?, ?, ?)
+"#;
+
+static LIST_KEYSPACES_QUERY: &str = r#"
+    SELECT keyspace_id, name, namespace, primary_zone, base_key_ranges
+    FROM chardonnay.keyspaces
+"#;
+
+// TODO: Similar to tx_state_store. We should move this to a common location.
+fn get_serial_query(query_text: impl Into<String>) -> Query {
+    let mut query = Query::new(query_text);
+    query.set_serial_consistency(Some(SerialConsistency::Serial));
+    query
+}
+
+// TODO: Similar to rangeserver and tx_state_store. We should move this to a
+// common location.
+fn scylla_query_error_to_storage_error(qe: QueryError) -> Error {
+    match qe {
+        QueryError::TimeoutError | QueryError::DbError(DbError::WriteTimeout { .. }, _) => {
+            Error::Timeout
+        }
+        _ => Error::InternalError(Arc::new(qe)),
+    }
+}
+
+// Wrapper struct for the KeyspaceInfo proto.
+// This is needed because the KeyspaceInfo proto is not directly convertible to
+// a Scylla row. Instead, we need to implement the FromRow trait.
+#[derive(Debug)]
+struct KeyspaceInfoWrapper(KeyspaceInfo);
+
+#[derive(Debug, FromUserType, SerializeValue)]
+struct SerializedRegion {
+    name: String,
+    cloud: Option<String>,
+}
+
+#[derive(Debug, FromUserType, SerializeValue)]
+struct SerializedZone {
+    region: Option<SerializedRegion>,
+    name: String,
+}
+
+#[derive(Debug, FromUserType, SerializeValue)]
+struct SerializedKeyRange {
+    base_range_uuid: Uuid,
+    lower_bound_inclusive: Vec<u8>,
+    upper_bound_exclusive: Vec<u8>,
+}
+
+#[derive(Debug, FromRow, SerializeRow)]
+struct SerializedKeyspaceInfo {
+    keyspace_id: Uuid,
+    name: String,
+    namespace: String,
+    primary_zone: Option<SerializedZone>,
+    // We need Option here because Scylla doesn't support empty lists.
+    // Without Option, on reading, deserialization will fail if the list is empty.
+    base_key_ranges: Option<Vec<SerializedKeyRange>>,
```

Comment:
There's no need to make this empty. If the request didn't include base key ranges, we can supply a default (i.e. one range that spans the entire keyspace)